libs/vkd3d/command.c

```diff
@@ -1839,6 +1839,120 @@ static void d3d12_command_list_invalidate_root_parameters(struct d3d12_command_l
     }
 }
 
+static void vk_access_and_stage_flags_from_d3d12_resource_state(const struct d3d12_device *device,
+        uint32_t state_mask, VkQueueFlags vk_queue_flags, VkPipelineStageFlags *stages, VkAccessFlags *access)
+{
+    VkPipelineStageFlags queue_shader_stages = 0;
+    uint32_t unhandled_state = 0;
+
+    if (vk_queue_flags & VK_QUEUE_GRAPHICS_BIT)
+    {
+        queue_shader_stages |= VK_PIPELINE_STAGE_VERTEX_SHADER_BIT |
+                VK_PIPELINE_STAGE_TESSELLATION_CONTROL_SHADER_BIT |
+                VK_PIPELINE_STAGE_TESSELLATION_EVALUATION_SHADER_BIT |
+                VK_PIPELINE_STAGE_GEOMETRY_SHADER_BIT |
+                VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT;
+    }
+
+    if (vk_queue_flags & VK_QUEUE_COMPUTE_BIT)
+        queue_shader_stages |= VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT;
+
+    if (state_mask == D3D12_RESOURCE_STATE_COMMON)
+    {
+        *stages |= VK_PIPELINE_STAGE_ALL_COMMANDS_BIT;
+        *access |= VK_ACCESS_MEMORY_READ_BIT | VK_ACCESS_MEMORY_WRITE_BIT;
+        return;
+    }
+
+    while (state_mask)
+    {
+        uint32_t state = state_mask & -state_mask;
+
+        switch (state)
+        {
+            case D3D12_RESOURCE_STATE_VERTEX_AND_CONSTANT_BUFFER:
+                *stages |= queue_shader_stages | VK_PIPELINE_STAGE_VERTEX_INPUT_BIT;
+                *access |= VK_ACCESS_UNIFORM_READ_BIT | VK_ACCESS_VERTEX_ATTRIBUTE_READ_BIT;
+
+                if (device->bindless_state.flags & VKD3D_BINDLESS_CBV_AS_SSBO)
+                    *access |= VK_ACCESS_SHADER_READ_BIT;
+                break;
+
+            case D3D12_RESOURCE_STATE_INDEX_BUFFER:
+                *stages |= VK_PIPELINE_STAGE_VERTEX_INPUT_BIT;
+                *access |= VK_ACCESS_INDEX_READ_BIT;
+                break;
+
+            case D3D12_RESOURCE_STATE_RENDER_TARGET:
+                *stages |= VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;
+                *access |= VK_ACCESS_COLOR_ATTACHMENT_READ_BIT | VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT;
+                break;
+
+            case D3D12_RESOURCE_STATE_UNORDERED_ACCESS:
+                *stages |= queue_shader_stages;
+                *access |= VK_ACCESS_SHADER_READ_BIT | VK_ACCESS_SHADER_WRITE_BIT;
+                break;
+
+            case D3D12_RESOURCE_STATE_DEPTH_WRITE:
+                *stages |= VK_PIPELINE_STAGE_EARLY_FRAGMENT_TESTS_BIT | VK_PIPELINE_STAGE_LATE_FRAGMENT_TESTS_BIT;
+                *access |= VK_ACCESS_DEPTH_STENCIL_ATTACHMENT_READ_BIT | VK_ACCESS_DEPTH_STENCIL_ATTACHMENT_WRITE_BIT;
```

Comment:
Actually, for the resource states where we do implicit state transitions anyways, I think we can use access of 0. We only need an execution barrier to link up the incoming barrier for render pass or transfer operation.

Reply:
E.g. for a state transition into RENDER_TARGET we will see two barrier:

SOMETHING -> RENDER_TARGET (flush something, block COLOR_ATTACHMENT)
BeginRenderPass -> here we wait for COLOR_ATTACHMENT, perform layout transition).

We never need memory to be visible before we perform the layout transition.

Reply:
Done.